imap_processing/lo/l0/science_direct_events.py

```diff
@@ -0,0 +1,202 @@
+        self.tof_calculation_binary = {
+            field: bitstring.Bits(calculation_value)
+            for field, calculation_value in tof_calculation_values.items()
+        }
+
+    def _find_decompression_case(self):
+        """Find the decompression case for this DE."""
+        # The first 4 bits of the binary data are used to
+        # determine which case number we are working with.
+        # The case number is used to determine how to
+        # decompress the TOF values.
+        self.case_number = ct.tof_case_table[self.DATA[0:4].bin]
+
+    def _find_remaining_bits(self):
+        """Find which TOF coefficients are needed for a data field and case number."""
+        self.remaining_bits = {}
+        for field, tof_calculation_binary in self.tof_calculation_binary.items():
+            # get a list of tof calculation values as integers
+            # We only need the last 12 bits from the tof calculation binary
+            tof_calculation_array = bitarray(tof_calculation_binary).tolist()[-12:]
+
+            # the tof calculation table binary are used to determine which values
+            # from the tof coefficient table should be used in combination with
+            # the tof decoder.
+            self.remaining_bits[field] = list(
+                compress(ct.tof_coefficient_table, tof_calculation_array)
+            )
+
+    def _parse_binary(self):
+        """Use the TOF decoder to split up the data bits into its TOF fields."""
+        # separate the binary data into its parts
+        self.parsed_bits = {}
+        # the first few bits are only used for determining the case number
+        # so either the first 5 or 6 bits need to be ignored depending the kind
+        # of DE.
+        if self.case_number in [0, 4, 6, 10, 12, 14]:
+            bit_start = 5
+        else:
+            bit_start = 4
+        # Use the TOF decoder to chunk the data binary into its componenets
+        # TOF0, TOF1, TOF2, etc.
+        for field, bit_length in self.tof_decoder.items():
+            self.parsed_bits[field] = self.DATA[bit_start : bit_start + bit_length]
+            bit_start = bit_start + bit_length
+
+    def _decode_fields(self):
+        """Use the parsed data and TOF coefficients to decode the binary."""
+        for field, tof_bits in self.parsed_bits.items():
+            needed_bits = tof_bits[-len(self.remaining_bits[field]) :]
+            # Use the TOF coefficients and the bits for the current field to
+            # calculate the decompressed value. Also round to 2 because the TOF
+            # coefficients only have 2 decimals.
+            decompressed_data = round(
+                sum(
+                    bit[0] * bit[1]
+                    for bit in zip(
+                        bitarray(needed_bits).tolist(), self.remaining_bits[field]
+                    )
+                ),
+                2,
+            )
+            setattr(self, field, decompressed_data)
```

Comment:
Does this error if `field` doesn't exist in the class? Is this overriding existing fields that were decommed (eg `self.TOF1`) or creating new ones? In general, I think `setattr` should be used with caution. It looks like `field` comes from `tof_decoder_table` and doesn't have that many values, so would it make sense to have TOF be a class with those fields and instead of using `setattr` here, you just create a new instance of that TOF class and then add it in here...? I'm not sure what the best way is, but I wanted to flag `setattr` to see if there's a more clear way (because I had to go back through several class variables to figure out what `field` would be)

Reply:
Yeah that makes sense. I figured it was safe because those fields are coming from the tables and not the data. Also, the TOF values and other attributes that are being set there were never decommed; they are being pulled from the binary only so they haven't been set yet. But I like the idea of creating a TOF class, that might be the way go.

Reply:
Generally, yes, having the fields come from the tables means that this is alright in my book, but it is a red flag that someone could change those tables later and not realize that they are used for setting the fields here. So, additional documentation might be good, or just making that relationship more explicit with code design such as by making it a namedtuple or a class.